[4.7.5] Implement reaction on ap_metric_interval

VladyslavTupikin opened the pull request:
agent: backhaul_manager: Implement reaction on ap_metric_interval

Certification test 4.7.5 sets ap_metric_interval as part of AP Policy Config message and the agent should send AP Metrics Response every ap_metric_interval seconds until an agent will get another AP Policy Config message with ap_metric_interval equal to 0.

Update the after_select method by adding part of the code for processing ap_metric_interval.

Commit:
agent: backhaul_manager: Create send_slave_ap_metric_query_message method

For the re-using the code, need to divide handle_ap_metrics_query method on
2 parts: first should process AP_METRICS_QUERY message and call new method,
second part which finds correct socket for the given BSSID should be moved
to the new method.

Create send_slave_ap_metric_query_message method and move code for the searching
correct socket from given BSSID and forwarding AP_METRICS_QUEY to the son_slave
from handle_ap_metrics_query to the send_slave_ap_metric_query_message.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -500,13 +500,37 @@ void backhaul_manager::after_select(bool timeout)
     }
 
+    // Send ap metrics query every reporting_interval_s seconds
+    static std::chrono::steady_clock::time_point ap_metric_interval =
```
There was a placeholder for this code in backhaul_fsm_main(), inside case EState::OPERATIONAL

VladyslavTupikin replied:
Fixed

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -500,13 +500,37 @@ void backhaul_manager::after_select(bool timeout)
+    elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - discovery_timestamp);
     if ((m_eFSMState == EState::CONNECTED) ||
-        (m_eFSMState == EState::OPERATIONAL && elapsed.count() == 60)) {
+        (m_eFSMState == EState::OPERATIONAL && elapsed.count() >= 60)) {
```
This fix should be in a separate commit, even in a separate PR

VladyslavTupikin replied:
Opened new issue https://github.com/prplfoundation/prplMesh/issues/1341

arnout replied:
It doesn't make sense to open an issue for splitting things in a separate PR or changing commits - once the PR is merged, it becomes irrelevant.

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -500,13 +500,37 @@ void backhaul_manager::after_select(bool timeout)
+            }
+        }
+
+        send_slave_ap_metrcic_query_message(0, bssid_list);
```
There is a typo in the name of the method

VladyslavTupikin replied:
fixed

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -500,13 +500,37 @@ void backhaul_manager::after_select(bool timeout)
+                }
+            }
+        }
+
```
Do not send message if bssid_list is empty.

VladyslavTupikin replied:
fixed

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2279,6 +2303,13 @@ bool backhaul_manager::handle_multi_ap_policy_config_request(ieee1905_1::CmduMes
         ap_metrics_reporting_info.reporting_interval_s =
             metric_reporting_policy_tlv->metrics_reporting_interval_sec();
         ap_metrics_reporting_info.last_reporting_time_point = std::chrono::steady_clock::now();
```
I think this is the equivalent to your static ap_metric_interval. If true, we should use one or the other, but not both. IMO, we should use member variables and not static variables.

mariomaz reviewed (changes requested):
I think there's some duplicate code.

Commit:
agent: backhaul_manager: Implement reaction on ap_metric_interval

Certification test 4.7.5 sets ap_metric_interval as part of AP Policy
Config message and the agent should send AP Metrics Response every
ap_metric_interval seconds until agent will get another AP Policy
Config message with ap_metric_interval equal to 0.

Update the backhaul_fsm_main method with adding part of code for processing
ap_metric_interval.

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

Commit:
Add 4.7.5_ETH_* tests to the list of passing tests.

MAP-4.7.5_ETH_FH24G:netgear-rax40
MAP-4.7.5_ETH_FH5GL:netgear-rax40
MAP-4.7.5_ETH_FH5GH:netgear-rax40

Signed-off-by: Vladyslav Tupikin <v.tupikin@inango-systems.com>

VladyslavTupikin commented:
This PR should be merged after https://github.com/prplfoundation/prplMesh/pull/1285

mariomaz commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -2302,6 +2317,13 @@ bool backhaul_manager::handle_multi_ap_policy_config_request(ieee1905_1::CmduMes
         ap_metrics_reporting_info.last_reporting_time_point = std::chrono::steady_clock::now();
+
+        if (ap_metrics_reporting_info.reporting_interval_s != 0) {
+            set_select_timeout(ap_metrics_reporting_info.reporting_interval_s *
```
I think that changing the select timeout will have side effects on the rest of actions that have to be executed periodically.

arnout replied:
Good point! The select timeout is currently set at a fixed 500ms. So there actually is no reason to change it since the reporting interval is going to be more than 1s. So @VladyslavTupikin please remove the set_select_timeout and add a comment that it's not needed since the timeout is 500ms.

VladyslavTupikin replied:
Fixed, added comment.

mariomaz reviewed (approved):
I think that changing the select timeout when reporting interval is set is affecting other periodic actions.
Nevertheless, since it is a very small change in case you agree with me, I'm approving.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1208,6 +1208,54 @@ bool backhaul_manager::send_autoconfig_search_message(std::shared_ptr<sRadioInfo
     LOG(DEBUG) << "sending autoconfig search message, bridge_mac=" << bridge_info.mac;
     return send_cmdu_to_bus(cmdu_tx, network_utils::MULTICAST_1905_MAC_ADDR, bridge_info.mac);
 }
+bool backhaul_manager::send_slave_ap_metric_query_message(uint16_t mid,
```
(nitpick) please leave an empty line before the function.

VladyslavTupikin replied:
fixed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1019,7 +1019,22 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
 
-                // TODO: report AP metrics
+                std::vector<sMacAddr> bssid_list;
+                for (auto &socket : slaves_sockets) {
```
```suggestion
                for (const auto &socket : slaves_sockets) {
```

VladyslavTupikin replied:
fixed

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1019,7 +1019,22 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
+                    }
+                }
+                if (!bssid_list.empty()) {
+                    send_slave_ap_metric_query_message(0, bssid_list);
```
I think this deserves a comment to explain the 0:
```suggestion
                    // We must generate a new MID for the response, so set MID to 0 here.
                    send_slave_ap_metric_query_message(0, bssid_list);
```

VladyslavTupikin replied:
Added

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:
```diff
@@ -1019,7 +1019,22 @@ bool backhaul_manager::backhaul_fsm_main(bool &skip_select)
+                    send_slave_ap_metric_query_message(0, bssid_list);
+                } else {
+                    LOG(DEBUG)
+                        << "Failed to send AP_METRICS_QUERY to the son slave, empty BSSID list";
```
It's not really a failure...
```suggestion
                        << "Skipping AP_METRICS_QUERY for slave " << socket->hostap_iface <<", empty BSSID list";
```

VladyslavTupikin replied:
Fixed just commentary, bssid_list contains all BSSID on all radio

arnout reviewed (approved):
Minor tweaks only.
